server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesDestructionProcessor.java

```diff
@@ -103,13 +86,15 @@ public SetMessagesResponse process(SetMessagesRequest request, MailboxSession ma
         };
     }
 
-    private MailboxMessage getMailboxMessage(MessageMapper messageMapper, MessageId messageId, Mailbox mailbox)
-            throws MailboxException, MessageNotFoundException {
-
-        Iterator<MailboxMessage> mailboxMessage = messageMapper.findInMailbox(mailbox, MessageRange.one(messageId.getUid()), MessageMapper.FetchType.Metadata, LIMIT_BY_ONE);
-        if (!mailboxMessage.hasNext()) {
+    private void checkThatMessageExists(MessageManager messageManager, MessageId messageId, MailboxSession mailboxSession) throws MailboxException, MessageNotFoundException {
+        MessageResultIterator messages = messageManager.getMessages(messageId.getUidAsRange(), FetchGroupImpl.MINIMAL, mailboxSession);
+        if (!messages.hasNext()) {
             throw new MessageNotFoundException();
         }
-        return mailboxMessage.next();
+    }
+
+    private void removeMessage(MessageManager messageManager, MessageId messageId, MailboxSession mailboxSession) throws MailboxException {
+        messageManager.setFlags(new Flags(Flags.Flag.DELETED), MessageManager.FlagsUpdateMode.ADD, messageId.getUidAsRange(), mailboxSession);
```

Comment:
If we do so, then we are implementing it at the manager level.
At the manager level we can deal with quotas, events, etc...
By the way if we right a delete method by uid then we can factorize it with the expunge method, reducing risks to break something.
It makes a lot of sens IMO, as it is what you are truely doing here. No need to keep lying to the underlying layers....